### PR TITLE
refactor(state_cache): simplify `StateCache`s by reusing `EstimatedBTreeMap`

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previously we have duplicated logic in `TopNCache`, `OrderedStateCache` and `EstimatedBTreeMap`. This PR reuses `EstimatedBTreeMap` in the former two.

This PR also removes `TopNCache` and embed the eviction logic after insertion into `TopNStateCache`, which is previously the only user or `TopNCache`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by kwannoel — APPROVED** [review]
LGTM, thanks!

File: src/common/estimate_size/src/collections/btreemap.rs [hunk 1]
```diff
@@ -62,20 +72,55 @@ where
     pub fn last_key_value(&self) -> Option<(&K, &V)> {
         self.inner.last_key_value()
     }
+}
+
+#[easy_ext::ext(BTreeMapAccessExt)]
```
Comment (xxchan) [inline src/common/estimate_size/src/collections/btreemap.rs:77]: Why to use easy_ext here?

Comment (stdrc) [reply src/common/estimate_size/src/collections/btreemap.rs:77]: Because these are not BTreeMap methods

**Review by xxchan — APPROVED** [review]
LGTM

File: src/common/estimate_size/src/collections/btreemap.rs (hunk 1 above)
Comment (xxchan) [reply src/common/estimate_size/src/collections/btreemap.rs:77]: For me, extension trait is for
1. prevent overriding default impl
2. impl methods for 3rd party types (without using a newtype)

It's a little overkill to me here

**Pull request merged**